logstash-core/src/main/java/org/logstash/ackedqueue/QueueUpgrade.java

```diff
@@ -0,0 +1,118 @@
+            try (final DirectoryStream<Path> files = Files.newDirectoryStream(path)) {
+                final Collection<File> pageFiles = StreamSupport.stream(
+                    files.spliterator(), false
+                ).filter(
+                    file -> PAGE_NAME_PATTERN.matcher(file.getFileName().toString()).matches()
+                ).map(Path::toFile).collect(Collectors.toList());
+                final CheckpointIO cpIo = new FileCheckpointIO(path);
+                pageFiles.forEach(p -> validatePageFile(path, cpIo, p));
+                pageFiles.forEach(QueueUpgrade::setV2);
+            }
+            Files.write(upgradeFile.toPath(), Ints.toByteArray(2), StandardOpenOption.CREATE);
+        }
+    }
+
+    private static void validatePageFile(final Path path, final CheckpointIO cpIo, final File v1PageFile) {
+        final int num =
+            Integer.parseInt(v1PageFile.getName().substring("page.".length()));
+        try (final MmapPageIOV1 iov1 = new MmapPageIOV1(
+            num, Ints.checkedCast(v1PageFile.length()), path
+        )) {
+            final Checkpoint cp = loadCheckpoint(path, cpIo, num);
+            final int count = cp.getElementCount();
+            final long minSeqNum = cp.getMinSeqNum();
+            iov1.open(minSeqNum, count);
+            for (int i = 0; i < count; ++i) {
+                verifyEvent(iov1, minSeqNum + i);
+            }
+        } catch (final IOException ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    private static void verifyEvent(final PageIO iov1, final long seqNum) {
+        try {
+            Event.deserialize(iov1.read(seqNum, 1).getElements().get(0));
+        } catch (final IOException ex) {
+            failValidation(ex);
+        }
+    }
+
+    private static void setV2(final File v1PageFile) {
+        try (final RandomAccessFile raf = new RandomAccessFile(v1PageFile, "rw")) {
+            raf.seek(0L);
+            raf.writeByte((int) MmapPageIOV2.VERSION_TWO);
+        } catch (final IOException ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    private static Checkpoint loadCheckpoint(final Path path, final CheckpointIO cpIo,
+        final int num) throws IOException {
+        final String cpFilename = cpIo.tailFileName(num);
+        final Checkpoint cp;
+        if (path.resolve(cpFilename).toFile().exists()) {
+            cp = cpIo.read(cpFilename);
+        } else {
+            cp = cpIo.read("checkpoint.head");
+            if (cp.getPageNum() != num) {
+                throw new IllegalStateException(
+                    String.format("No checkpoint file found for page %d", num)
+                );
+            }
+        }
+        return cp;
+    }
+
+    private static void failValidation(final Throwable ex) {
+        LOGGER.error("Logstash was unable to upgrade your persistent queue data." +
+            "Please either downgrade to version 6.2.3 and fully drain " +
```

Comment:
Just wondering, is the recommendation to downgrade to 6.2.3 or "to your previous Logstash version" better? Sorry if you've already discussed this.

Reply:
Yes, we should do the previous Logstash version. Armin and I already agreed on this in a discussion